internal/grpc/xds.go

```diff
@@ -117,10 +118,10 @@ func (xh *xdsHandler) stream(st grpcStream) (err error) {
 				}
 
 				resp := &v2.DiscoveryResponse{
-					VersionInfo: "0",
+					VersionInfo: strconv.Itoa(last),
```

Comment:
I spent the day looking at protocol too and came to a the same conclusion (exposing last as the version and nonce). I had gone a slightly different route, altering the Cache interface function values to return both the ` ([]proto.Message, int)`. I like yours more; it is a lot simpler :).

If we walk through this a little:
 - Update(v map[string]*v2.RouteConfiguration) : lock, swap values and notifies (`last` on channel), unlock
 - read `last` from channel
 - Values(func(string) bool) ([]proto.Message, int): lock, extract/filter values, unlock
 - toAny(r, toFilter(req.ResourceNames)): `proto.Marshal` transform
 - st.Send(resp): send

There appears to be a small window between reading `last` from the channel and the lock to read values where change could occur, that would result in sending the same set of values for two  sequence numbers. The window expands as the time to `proto.Marshal` transform and `st.Send(resp)` expands, as it might as data sizes increase.

For example:
 - while creating and processing `st.Send(resp)`
 - Update(v map[string]*v2.RouteConfiguration): lock, `last` on channel, unlock
 - Update(v map[string]*v2.RouteConfiguration): lock, buffered channel blocks
 - read `last` from channel
 -  Update(v map[string]*v2.RouteConfiguration): owns lock, `last` on channel, unlock
 - toAny(r, toFilter(req.ResourceNames)): lock, extract/filter values, unlock
 - toAny(r, toFilter(req.ResourceNames)): lock, extract/filter values, unlock
> these two sends provide the same set of data at different sequence numbers, will this violate the "When to send an update" condition of the xDS api?

If I missed something there, I apologize.  Getting version number in here is good regardless.

Reply:
> Each stream has its own notion of versioning, there is no shared versioning across resource types. When ADS is not used, even each resource of a given resource type may have a distinct version

That seems important. Is `last` shared?

Reply:
`last` is tracked per-cache (acts as a cache sequence number), you'll see a separate value in `ClusterCache`, `ListenerCache`, and `RouteCache`.

Reply:
I'll go run my handy-dandy "create a bunch of ingresses" script to see if this alters memory behavior.